libs/qCC_db/src/ccRasterGrid.cpp

```diff
@@ -401,79 +292,197 @@ bool ccRasterGrid::fillWith(	ccGenericPointCloud* cloud,
-					if (hasColors && projectionType == PROJ_AVERAGE_VALUE)
-					{
-						cell.color /= cell.nbPoints;
+					assert(cloud->hasColors());
+					if (projectionType == PROJ_AVERAGE_VALUE){
+						//Average Color
+						aCell.color = CCVector3d(0,0,0);
+						for (unsigned n=0; n < aCell.nbPoints ;n++){
+							unsigned pointIndex = cellPointIndexedHeight[n].index;
+							const ccColor::Rgb& col = cloud->getPointColor(pointIndex);
+							aCell.color += CCVector3d(col.r, col.g, col.b);
+						}
+						aCell.color /= aCell.nbPoints;
+					}
+					else{  
+						//Pick color from selected index
+						const ccColor::Rgb& col = cloud->getPointColor(aCell.pointIndex);
+						aCell.color = CCVector3d(col.r, col.g, col.b);
 					}
 				}
-				else
-				{
-					cell.stdDevHeight = 0;
-				}
-
-				if (cell.nbPoints != 0)
+				
+				//Handle scalar fields
+				if (interpolateSF)
 				{
-					//set the right 'height' value
-					switch (projectionType)
+					assert(pc);
+					//absolute position of the cell (e.g. in the 2D SF grid(s))
+					int pos = cellPos.y * static_cast<int>(width) + cellPos.x;
+					assert(pos < static_cast<int>(gridTotalSize));
+					
+					for (size_t k = 0; k < scalarFields.size(); ++k)
 					{
-					case PROJ_MINIMUM_VALUE:
-						cell.h = cell.minHeight;
-						break;
-					case PROJ_AVERAGE_VALUE:
-						cell.h = cell.avgHeight;
-						break;
-					case PROJ_MAXIMUM_VALUE:
-						cell.h = cell.maxHeight;
-						break;
-					default:
-						assert(false);
-						break;
+						assert(!scalarFields[k].empty());
+						CCCoreLib::ScalarField* sf = pc->getScalarField(static_cast<unsigned>(k));
+						assert(sf && pos < scalarFields[k].size());
+						// Set up vector of valid sf values for current cell 
+						cellPointSF.reserve(aCell.nbPoints);
+						unsigned validPoints=0;
+						for (unsigned n=0; n < aCell.nbPoints ;n++){
+							unsigned pointIndex = cellPointIndexedHeight[0].index;
+							ScalarType sfValue = sf->getValue(pointIndex);
+							if(std::isfinite(sfValue)){
+								cellPointSF[validPoints] = sfValue;
+								validPoints++;
+							}
+						}
+						if (validPoints==0) continue;
+						auto cellPointSF_end = std::next(cellPointSF.begin(), validPoints);
+						
+						// Sort valid scalar values for cell
+						std::sort(cellPointSF.begin(), cellPointSF_end, [](ScalarType a, ScalarType b) {
+							return a<b;
+						});
+
+						switch (sfInterpolation)
+						{
+							case PROJ_MINIMUM_VALUE:
+								scalarFields[k][pos] = cellPointSF[0];
+								break;
+							case PROJ_AVERAGE_VALUE:
+								scalarFields[k][pos] = std::accumulate(cellPointSF.begin(), cellPointSF_end, 0.0) / validPoints;
+								break;
+							case PROJ_MEDIAN_VALUE:
+								if(validPoints%2){
+									scalarFields[k][pos] = cellPointSF[(validPoints/2)];
+								}
+								else{
+									scalarFields[k][pos] = (cellPointSF[(validPoints/2)-1] + cellPointSF[(validPoints/2)])/2;
```

Comment:
same comment as above if validPoints == 1

Reply:
Same as above. The line in question will only run if n is even and non-zero